lib/stitches/running_stitch.py

```diff
@@ -49,69 +51,189 @@ def split_segment_random_phase(a, b, length: float, length_sigma: float, random_
+            return None
+        segArc = AngleInterval.fromSegment(a - origin, b - origin)
+        if segArc is None:
+            return a  # b is exactly behind origin from a
+        if segArc.containsAngle(self.a):
+            return cut_segment_with_angle(origin, self.a, a, b)
+        elif segArc.containsAngle(self.b):
+            return cut_segment_with_angle(origin, self.b, a, b)
+        else:
+            return None
+
+
+def cut_segment_with_angle(origin: Point, angle: float, a: Point, b: Point) -> Point:
+    # Assumes the crossing is inside the segment
+    p = a - origin
+    d = b - a
+    c = Point(math.cos(angle), math.sin(angle))
+    t = (p.y*c.x - p.x*c.y) / (d.x*c.y - d.y*c.x)
+    if t < -0.000001 or t > 1.000001:
+        raise Exception("cut_segment_with_angle returned a parameter of {0} with points {1} {2} and cut line {3} ".format(t, p, b-origin, c))
+    return a + d*t
+
+
+def cut_segment_with_circle(origin: Point, r: float, a: Point, b: Point) -> Point:
+    # assumes that a is inside the circle and b is outside
+    p = a - origin
+    d = b - a
+    # inner products
+    p2 = p * p
+    d2 = d * d
+    r2 = r * r
+    pd = p * d
+    # r2 = p2 + 2*pd*t + d2*t*t, quadratic formula
+    t = (math.sqrt(pd*pd + r2*d2 - p2*d2) - pd) / d2
+    if t < -0.000001 or t > 1.000001:
+        raise Exception("cut_segment_with_circle returned a parameter of {0}".format(t))
+    return a + d*t
+
+
+def take_stitch(start: Point, points: typing.Sequence[Point], idx: int, stitch_length: float, tolerance: float):
+    # Based on a single step of the Zhao-Saalfeld curve simplification algorithm.
+    # https://cartogis.org/docs/proceedings/archive/auto-carto-13/pdf/linear-time-sleeve-fitting-polyline-simplification-algorithms.pdf
+    # Adds early termination condition based on stitch length.
+    if idx >= len(points):
+        return None, None
+
+    sleeve = AngleInterval.all()
+    last = start
+    for i in range(idx, len(points)):
+        p = points[i]
+        if sleeve.containsPoint(p - start):
+            if start.distance(p) < stitch_length:
+                sleeve = sleeve.intersect(AngleInterval.fromBall(p - start, tolerance))
+                last = p
+                continue
+            else:
+                cut = cut_segment_with_circle(start, stitch_length, last, p)
+                return cut, i
+        else:
+            cut = sleeve.cutSegment(start, last, p)
+            if start.distance(cut) > stitch_length:
+                cut = cut_segment_with_circle(start, stitch_length, last, p)
+            return cut, i
+    return points[-1], None
+
+
+def stitch_curve_even(points: typing.Sequence[Point], stitch_length: float, tolerance: float):
+    # includes end point but not start point.
     if len(points) < 2:
         return []
+    distLeft = [0] * len(points)
+    for i in reversed(range(0, len(points) - 1)):
+        distLeft[i] = distLeft[i + 1] + points[i].distance(points[i+1])
```

Comment:
I think it's possible to do this with numpy using just one or two function calls.  Might be faster?

Reply:
Numpy doesn't take Point objects.